### PR TITLE
Add town buy and sell feature

### PR DESCRIPTION
**Pull request by Fruitloopins**
#### Description: 

This pull request adds a few commands relating to selling and buying a town, there are confirmations for putting your town up for sale and for buying a town, including a warning that it will remove you from your current town. This is my first time making a pull request for Towny so I'm sure I've missed an important edge case or something but I tried to test it as thoroughly as I can with my understanding of Towny. I did not add the `/t list by sale` functionality from the original issue #6478 yet.

____
#### New Nodes/Commands/ConfigOptions: 

- `/t forsale/fs [$]` Puts town up for sale at the specified amount
  - `towny.command.town.forsale`
- `/t notforsale/nfs` Cancels town sale
  - `towny.command.town.notforsale`
- `/t buytown [town]` Buy a town that is for sale
  - `towny.command.town.buytown`
- `/ta town [town] forsale/fs [$]` Admin command to force town sale at specified price
  - `towny.command.townyadmin.town.forsale`
- `/ta town [town] notforsale/nfs` Admin command to force cancel town sale
  - `towny.command.townyadmin.town.notforsale`

____
#### Relevant Towny Issue ticket:

#6478 

____
- [x] I have tested this pull request for defects on a server. 

By making this pull request, I represent that I have the right to waive copyright and related rights to my contribution, and agree that all copyright and related rights in my contributions are waived, and I acknowledge that the TownyAdvanced organization has the copyright to use and modify my contribution under the Towny [License](https://github.com/LlmDl/Towny/blob/master/LICENSE.md) for perpetuity.

### CODE REVIEW

File: Towny/src/main/java/com/palmergames/bukkit/towny/command/TownCommand.java [hunk 1]
```diff
@@ -4459,4 +4479,85 @@ private static int homeBlockDistance(Town town1, Town town2) {
...
+		}
+
+		if (!town.isForSale()) {
+			throw new TownyException(Translatable.of("msg_town_buytown_not_forsale"));
+		}
+
+		Resident resident = getResidentOrThrow(player);
+		if (town.getMayor() == resident) {
+			throw new TownyException(Translatable.of("msg_town_buytown_already_mayor", town.getName()));
+		}
```
Comment (LlmDl) [inline Towny/src/main/java/com/palmergames/bukkit/towny/command/TownCommand.java:4544]:
```suggestion
		if (resident.isMayor()) {
			throw new TownyException(Translatable.of("msg_town_buytown_already_mayor", resident.getTownOrNull().getName()));
		}

```

I think it ought to stop any mayor from buying a town, as any mayor that buys a town will stop being a mayor of their own town.

This suggestion above is not as good as having something more descriptive when `town != resident.getTownOrNull()`, which I think should explain that the player has to set a new mayor before they can buy another town. Similar to when a mayor does `/t leave`.

File: Towny/src/main/java/com/palmergames/bukkit/towny/command/TownCommand.java [hunk 2]
```diff
@@ -4459,4 +4479,85 @@ private static int homeBlockDistance(Town town1, Town town2) {
...
+						resident.removeTown();
+					townAddResident(town, resident);
+					town.setMayor(resident);
+				} catch (AlreadyRegisteredException e) {
+					town.setMayor(resident);
+				}
+				TownyMessaging.sendPrefixedTownMessage(town, Translatable.of("msg_new_mayor", resident.getName()));
+				town.setForSale(false);
+			})
+			.setTitle(Translatable.of("msg_town_buytown_confirmation", town.getName(), TownyEconomyHandler.getFormattedBalance(town.getForSalePrice())))
```
Comment (LlmDl) [inline Towny/src/main/java/com/palmergames/bukkit/towny/command/TownCommand.java:4559]:
```suggestion
			.setTitle(Translatable.of("msg_town_buytown_confirmation", town.getName(), prettyMoney(town.getForSalePrice())))

```

File: Towny/src/main/java/com/palmergames/bukkit/towny/command/TownyAdminCommand.java [hunk 3]
```diff
@@ -1313,6 +1320,16 @@ public void parseAdminTownCommand(CommandSender sender, String[] split) throws T
...
 			TownCommand.townBuy(sender, StringMgmt.remArgs(split, 2), town, true);
 			break;
+		case "forsale":
+		case "fs":
+			checkPermOrThrow(sender, TOWNY_COMMAND_TOWNYADMIN_TOWN_FORSALE.getNode());
+			parseAdminTownForSaleCommand(sender, town, StringMgmt.remArgs(split, 2));
+			break;
+		case "notforsale":
+		case "nfs":
+			checkPermOrThrow(sender, TOWNY_COMMAND_TOWNYADMIN_TOWN_NOTFORSALE.getNode());
```
Comment (LlmDl) [inline Towny/src/main/java/com/palmergames/bukkit/towny/command/TownyAdminCommand.java:1330]:
```suggestion
			checkPermOrThrow(sender, PermissionNodes.TOWNY_COMMAND_TOWNYADMIN_TOWN_NOTFORSALE.getNode());

```

File: Towny/src/main/java/com/palmergames/bukkit/towny/command/TownyAdminCommand.java [hunk 4]
```diff
@@ -2493,4 +2510,20 @@ private void parseAdminInstall(CommandSender sender) throws NoPermissionExceptio
...
+		double forSalePrice = Double.parseDouble(arg[0]);
+		TownCommand.setTownForSale(town, forSalePrice, true);
+
+		TownyMessaging.sendMsg(sender, Translatable.of("msg_town_forsale", town.getName(), TownyEconomyHandler.getFormattedBalance(forSalePrice)));
+	}
+
+	private void parseAdminTownNotForSaleCommand(CommandSender sender, Town town) {
+		TownCommand.setTownNotForSale(town, true);
+
+		TownyMessaging.sendMsg(sender, Translatable.of("msg_town_notforsale", town.getName()));
```
Comment (LlmDl) [inline Towny/src/main/java/com/palmergames/bukkit/towny/command/TownyAdminCommand.java:2527]: It might be a good idea to also send a town message, so the town members know they are not for sale anymore.

File: Towny/src/main/java/com/palmergames/bukkit/towny/command/TownyAdminCommand.java [hunk 5]
```diff
@@ -2493,4 +2510,20 @@ private void parseAdminInstall(CommandSender sender) throws NoPermissionExceptio
...
+			throw new TownyException(Translatable.of("msg_error_must_be_num"));
+
+		double forSalePrice = Double.parseDouble(arg[0]);
+		TownCommand.setTownForSale(town, forSalePrice, true);
+
+		TownyMessaging.sendMsg(sender, Translatable.of("msg_town_forsale", town.getName(), TownyEconomyHandler.getFormattedBalance(forSalePrice)));
+	}
+
+	private void parseAdminTownNotForSaleCommand(CommandSender sender, Town town) {
+		TownCommand.setTownNotForSale(town, true);
```
Comment (LlmDl) [inline Towny/src/main/java/com/palmergames/bukkit/towny/command/TownyAdminCommand.java:2525]: Before this, there should be logic for when a town is not for sale, with similar feedback to the TownCommand version.

File: Towny/src/main/java/com/palmergames/bukkit/towny/db/TownySQLSource.java
Comment (LlmDl) [inline Towny/src/main/java/com/palmergames/bukkit/towny/db/TownySQLSource.java:1]: You also need to add two columns into the SQLSChema.java class for this class to have the columns to work with.

File: Towny/src/main/java/com/palmergames/bukkit/towny/db/TownySQLSource.java [hunk 6]
```diff
@@ -1005,6 +1005,8 @@ private boolean loadTown(ResultSet rs) {
 			town.setAllowedToWar(rs.getBoolean("allowedToWar"));
 			town.setJoinedNationAt(rs.getLong("joinedNationAt"));
 			town.setMovedHomeBlockAt(rs.getLong("movedHomeBlockAt"));
+			town.setForSale(rs.getBoolean("forSale"));
```
Comment (LlmDl) [inline Towny/src/main/java/com/palmergames/bukkit/towny/db/TownySQLSource.java:1008]: These two new options will not appear and I believe it will be safer to do what is being done for the maxPercentTaxAmount:
line = rs.getBoolean("forSale")
if (line != null)
    town.setForSale(line);

This should be done for both forSale and forSalePrice.

File: Towny/src/main/java/com/palmergames/bukkit/towny/object/Government.java
Comment (LlmDl) [inline Towny/src/main/java/com/palmergames/bukkit/towny/object/Government.java:1]: All of the changes made to Government should be moved to Town. We don't want these methods accessible to Nations.

File: Towny/src/main/resources/lang/en-US.yml [hunk 7]
```diff
@@ -42,7 +42,7 @@ description: >
 # and formatting. Learn about MiniMessage formatting here:
 # https://docs.adventure.kyori.net/minimessage/format.html
 #
-# 
+#
```
Comment (LlmDl) [inline Towny/src/main/resources/lang/en-US.yml:45]: Unneeded diff.

File: Towny/src/main/java/com/palmergames/bukkit/towny/command/HelpMenu.java [hunk 8]
```diff
@@ -558,7 +560,6 @@ protected MenuBuilder load() {
 				.add("[days]", "");
 		}
 	},
-
```
Comment (LlmDl) [inline Towny/src/main/java/com/palmergames/bukkit/towny/command/HelpMenu.java:561]: Unneeded diff.

**Review by LlmDl — CHANGES_REQUESTED** [review]

File: Towny/src/main/java/com/palmergames/bukkit/towny/command/HelpMenu.java [hunk 9]
```diff
@@ -432,6 +432,8 @@ protected MenuBuilder load() {
 				.add("outlaw add/remove [name]", "")
 				.add("say", "[message]")
 				.add("spawn", Translation.of("town_help_5"))
+				.add("forsale [$]", "")
+				.add("notforsale [$]", "")
```
Comment (LlmDl) [inline Towny/src/main/java/com/palmergames/bukkit/towny/command/HelpMenu.java:436]: buytown should be added here as well

File: Towny/src/main/java/com/palmergames/bukkit/towny/command/TownCommand.java [hunk 10]
```diff
@@ -4459,4 +4495,88 @@ private static int homeBlockDistance(Town town1, Town town2) {
...
+						resident.removeTown();
+					townAddResident(town, resident);
+					town.setMayor(resident);
+				} catch (AlreadyRegisteredException e) {
+					town.setMayor(resident);
+				}
+				
+				TownyMessaging.sendPrefixedTownMessage(town, Translatable.of("msg_new_mayor", resident.getName()));
+				
+				currentMayor.getAccount().deposit(town.getForSalePrice(), "Payment for town sale");
```
Comment (LlmDl) [inline Towny/src/main/java/com/palmergames/bukkit/towny/command/TownCommand.java:4573]: There is always a possibility that a town's mayor is null, you should only pay if currentMayor != null

File: Towny/src/main/java/com/palmergames/bukkit/towny/db/SQLSchema.java [hunk 11]
```diff
@@ -240,6 +240,8 @@ private static List<String> getTownColumns() {
 		columns.add("`enemies` mediumtext NOT NULL");
 		columns.add("`hasUnlimitedClaims` bool NOT NULL DEFAULT '0'");
 		columns.add("`manualTownLevel` BIGINT DEFAULT '-1'");
+		columns.add("'forSale' bool NOT NULL DEFAULT '0'");
+		columns.add("'forSalePrice' float NOT NULL");
```
Comment (LlmDl) [inline Towny/src/main/java/com/palmergames/bukkit/towny/db/SQLSchema.java:244]:
```suggestion
		columns.add("`forSale` bool NOT NULL DEFAULT '0'");
		columns.add("`forSalePrice` float NOT NULL");

```

File: Towny/src/main/resources/plugin.yml [hunk 12]
```diff
@@ -353,6 +353,7 @@ permissions:
             towny.command.town.claim.*: true
             towny.command.town.delete: true
             towny.command.town.deposit: true
+            towny.command.town.forsale.*: true
```
Comment (LlmDl) [inline Towny/src/main/resources/plugin.yml:356]: These should not be 1 node with three children, it should be three nodes.

File: Towny/src/main/resources/plugin.yml [hunk 13]
```diff
@@ -384,6 +385,14 @@ permissions:
...
             towny.command.town.trusttown: true
 
+    towny.command.town.forsale.*:
+        Description: Users are able to use commands relating to buying and selling towns.
+        default: false
+        children:
+            towny.command.town.buytown: true
+            towny.command.town.forsale: true
+            towny.command.town.notforsale: true
+
```
Comment (LlmDl) [inline Towny/src/main/resources/plugin.yml:395]: Remove this

**Review by LlmDl — CHANGES_REQUESTED** [review]

**Review by LlmDl — CHANGES_REQUESTED** [review]

File: Towny/src/main/java/com/palmergames/bukkit/towny/command/TownCommand.java [hunk 14]
```diff
@@ -4459,4 +4495,95 @@ private static int homeBlockDistance(Town town1, Town town2) {
...
+		
+		Resident resident = getResidentOrThrow(player);
+		if (resident.isMayor()) {
+			throw new TownyException(Translatable.of("msg_town_buytown_already_mayor", resident.getTownOrNull().getName()));
+		}
+
+		Confirmation
+			.runOnAccept(() -> {
+				Resident currentMayor = town.getMayor();
+
```
Comment (LlmDl) [inline Towny/src/main/java/com/palmergames/bukkit/towny/command/TownCommand.java:4562]: This is a scenario where we shouldn't be using the `Confirmation.setCost(new ConfirmationTransaction(town::getForSalePrice, resident.getAccount(), "Town purchase cost."))` but instead use a more traditional money-check-and-pay in the Confirmation runnable to improve readability and prevent a resident paying when they cannot.

The runnable should go through three steps:
1. Test of the town is still for sale.
2. Test if the resident is already mayor.
3. Test if the resident can pay.
```java
if (!resident.getAccount().withdraw(cost)) {
    TownyMessaging.sendErrorMessage(resident, Translatable.of("YOU CANNOT AFFORD THIS"); <--- en-US.yml will have a string already in it.
    return;
}
```
Then it should begin to change the mayor over and do the rest of the work.


Afterwards, I would like you to test this out with teh following steps:

1. Resident A (mayor) sets the town for sale for $100.
5. Resident B uses the buytown command but does not confirm.
6. Resident A sets Resident B to be the mayor manually using /t set mayor.
7. Resident B hits the confirm.

File: Towny/src/main/java/com/palmergames/bukkit/towny/command/HelpMenu.java (hunk 8 above)
Comment (LlmDl) [reply Towny/src/main/java/com/palmergames/bukkit/towny/command/HelpMenu.java:561]: Still an unneeded diff.

**Review by LlmDl — CHANGES_REQUESTED** [review]

File: Towny/src/main/java/com/palmergames/bukkit/towny/TownyFormatter.java [hunk 15]
```diff
@@ -837,6 +837,8 @@ public static List<String> getTownSubtitle(Town town, TownyWorld world, Translat
 			sub.add(translator.of("status_town_title_peaceful"));
 		if (town.isConquered())
 			sub.add(translator.of("msg_conquered"));
+		if (town.isForSale())
+			sub.add(translator.of("status_forsale", TownyEconomyHandler.getFormattedBalance(town.getForSalePrice())));
```
Comment (LlmDl) [inline Towny/src/main/java/com/palmergames/bukkit/towny/TownyFormatter.java:841]:
```suggestion
			sub.add(translator.of("status_forsale", formatMoney(town.getForSalePrice())));

```

**Review by LlmDl — CHANGES_REQUESTED** [review]

File: Towny/src/main/java/com/palmergames/bukkit/towny/command/TownCommand.java (hunk 14 above)
Comment (Fruitloopins) [reply Towny/src/main/java/com/palmergames/bukkit/towny/command/TownCommand.java:4562]: from your test

image

File: Towny/src/main/java/com/palmergames/bukkit/towny/command/TownCommand.java [hunk 16]
```diff
@@ -526,6 +531,22 @@ public List<String> onTabComplete(CommandSender sender, Command command, String
...
+						for (Town town : TownyUniverse.getInstance().getTowns()) {
+							if (!town.isForSale()) {
+								notForSale.add(town.getName());
+							}
+						}
+
+						townsList.removeAll(notForSale);
+						return townsList;
+					}
```
Comment (LlmDl) [inline Towny/src/main/java/com/palmergames/bukkit/towny/command/TownCommand.java:548]:
```suggestion
				case "buytown":
					if (args.length == 2) {
						List<String> townsList = getTownyStartingWith(args[1], "t");
						townsList.removeIf(n -> !TownyAPI.getInstance().getTown(n).isForSale());
						return townsList;
					}
```

**Review by LlmDl — CHANGES_REQUESTED** [review]

**Review by LlmDl — APPROVED** [review]
